internal/k8s/replicaset/replicaset.go

```diff
@@ -0,0 +1,161 @@
+//
+// Copyright (C) 2019-2020 Vdaas.org Vald team ( kpango, rinx, kmrmt )
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package replicaset
+
+import (
+	"context"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/k8s"
+	"github.com/vdaas/vald/internal/log"
+)
+
+// ReplicaSetWatcher is a type alias for k8s resource controller.
+type ReplicaSetWatcher k8s.ResourceController
+
+type reconciler struct {
+	ctx         context.Context
+	mgr         manager.Manager
+	name        string
+	namespace   string
+	onError     func(err error)
+	onReconcile func(rs map[string][]ReplicaSet)
+	pool        sync.Pool
+}
+
+// ReplicaSet is a type alias for the k8s replica set definition.
+type ReplicaSet = appsv1.ReplicaSet
+
+// New returns the ReplicaSetWatcher that implements reconciliation loop, or any error occurred.
+func New(opts ...Option) (ReplicaSetWatcher, error) {
+	r := &reconciler{
+		pool: sync.Pool{
+			New: func() interface{} {
+				return make(map[string][]ReplicaSet)
+			},
+		},
+	}
+
+	for _, opt := range opts {
+		if err := opt(r); err != nil {
+			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
+		}
+	}
+
+	return r, nil
+}
+
+// Reconcile implements k8s reconciliation loop to retrieve the ReplicaSet information from k8s.
+func (r *reconciler) Reconcile(req reconcile.Request) (res reconcile.Result, err error) {
+	rsl := new(appsv1.ReplicaSetList)
+
+	err = r.mgr.GetClient().List(r.ctx, rsl)
+	if err != nil {
+		if r.onError != nil {
+			r.onError(err)
+		}
+		res = reconcile.Result{
+			Requeue:      true,
+			RequeueAfter: time.Millisecond * 100,
+		}
+		if k8serrors.IsNotFound(err) {
+			log.Warn(errors.ErrK8sResourceNotFound(err))
+			res.RequeueAfter = time.Second
+			return res, nil
+		}
+		return
+	}
+
+	// reset the last result cache
+	lrr := r.pool.Get().(map[string][]ReplicaSet)
+	for name := range lrr {
+		lrr[name] = lrr[name][:0]
+	}
```

Comment:
this code shoud be line 123

Reply:
you mean put the empty slice into pool, right?

Reply:
yes, we should store zero seeked slice, it has enough capacity.
we don't need to allocate next time.

Reply:
and your code also initialize after pool.Get, it means same.

Reply:
well...So my understanding is that we do not have to store the current reconcile result in the pool because of no using the data anymore (we need to know only the capacity of slices), is this correct along with your comment?

Reply:
Yes, the pool should store zero-seek slices with cap allocations, and there is no need to store the data.

Reply:
thx. I got it clearly 😎